Zend/zend_operators.c

```diff
@@ -2691,6 +2709,34 @@ static zend_always_inline void zend_str_tolower_impl(char *dest, const char *str
 }
 /* }}} */
 
+static zend_always_inline void zend_str_toupper_impl(char *dest, const char *str, size_t length) /* {{{ */ {
+	unsigned char *p = (unsigned char*)str;
+	unsigned char *q = (unsigned char*)dest;
+	unsigned char *end = p + length;
+#ifdef __SSE2__
+	if (length >= 16) {
+		const __m128i _a = _mm_set1_epi8('a' - 1);
+		const __m128i z_ = _mm_set1_epi8('z' + 1);
+		const __m128i delta = _mm_set1_epi8('a' - 'A');
+		do {
+			__m128i op = _mm_loadu_si128((__m128i*)p);
+			__m128i gt = _mm_cmpgt_epi8(op, _a);
+			__m128i lt = _mm_cmplt_epi8(op, z_);
+			__m128i mingle = _mm_and_si128(gt, lt);
+			__m128i sub = _mm_and_si128(mingle, delta);
+			__m128i upper = _mm_sub_epi8(op, sub);
+			_mm_storeu_si128((__m128i *)q, upper);
+			p += 16;
+			q += 16;
+		} while (p + 16 <= end);
```

Comment:
We can improve slightly on the `p + 16 <= end` by defining `z` to be `end` rounded down to a multiple of 16 and then simply compare `p != z`. Like this, if I haven't made a mistake (I'm quite tired, so likely):

```suggestion
		// round end down to multiple of 16
		static const uintptr_t mask = 15u;
		unsigned char *z = (unsigned char*)(((uintptr_t)end) & ~mask);
		const __m128i _a = _mm_set1_epi8('a' - 1);
		const __m128i z_ = _mm_set1_epi8('z' + 1);
		const __m128i delta = _mm_set1_epi8('a' - 'A');
		do {
			__m128i op = _mm_loadu_si128((__m128i*)p);
			__m128i gt = _mm_cmpgt_epi8(op, _a);
			__m128i lt = _mm_cmplt_epi8(op, z_);
			__m128i mingle = _mm_and_si128(gt, lt);
			__m128i sub = _mm_and_si128(mingle, delta);
			__m128i upper = _mm_sub_epi8(op, sub);
			_mm_storeu_si128((__m128i *)q, upper);
			p += 16;
			q += 16;
		} while (p != z);
```

Reply:
The assembly of the inner loop at -O2 is just

```
.L535:
	movdqu	-16(%rbx), %xmm1
	addq	$16, %rbx
	addq	$16, %rax
	movdqa	%xmm1, %xmm2
	movdqa	%xmm1, %xmm0
	pcmpgtb	%xmm4, %xmm2
	pcmpgtb	%xmm5, %xmm0
	pandn	%xmm6, %xmm2
	pand	%xmm2, %xmm0
	pand	%xmm3, %xmm0
	paddb	%xmm1, %xmm0
	movups	%xmm0, -16(%rax)
	cmpq	%rbx, %rbp
	jnb	.L535
```

%rbx holds p+16 which it uses for the loop comparison, and then it loads from %rbx-16. I don't think there's any overhead associated with the -16 offset -- my understanding is that the addressing modes all happen in the same clock cycle.

It's not correct to mask out the lower bits of the end pointer as you suggest, because the input is not aligned, hence the use of the loadu builtin.